usr/lib/pkcs11/ica_s390_stdll/ica_specific.c

```diff
@@ -118,822 +132,807 @@ ecc_support_in_libica_available(void)
-
-   if (rc != 0) {
-         TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
-         rc = CKR_FUNCTION_FAILED;
-   }
-   return rc;
+    CK_RV rc;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_KEY_TYPE keytype;
+    CK_BYTE key_value[3 * DES_KEY_SIZE];
+
+    // get the key type
+    rc = template_attribute_find(key->template, CKA_KEY_TYPE, &attr);
+    if (rc == FALSE) {
+        TRACE_ERROR("Could not find CKA_KEY_TYPE for the key.\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    keytype = *(CK_KEY_TYPE *) attr->pValue;
+
+    // get the key value
+    if (template_attribute_find(key->template, CKA_VALUE, &attr) == FALSE) {
+        TRACE_ERROR("Could not find CKA_VALUE for the key.\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    if (keytype == CKK_DES2) {
+        memcpy(key_value, attr->pValue, 2 * DES_KEY_SIZE);
+        memcpy(key_value + (2 * DES_KEY_SIZE), attr->pValue, DES_KEY_SIZE);
+    } else {
+        memcpy(key_value, attr->pValue, 3 * DES_KEY_SIZE);
+    }
+
+    rc = ica_3des_cmac_intermediate(message, (unsigned long) message_len,
+                                    (unsigned char *) key_value, mac);
+
+    if (rc != 0) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        rc = CKR_FUNCTION_FAILED;
+    }
+
+    return rc;
 }
 
 /*
  * Init SHA data structures
  */
 CK_RV token_specific_sha_init(STDLL_TokData_t *tokdata, DIGEST_CONTEXT *ctx,
-			      CK_MECHANISM *mech)
+                              CK_MECHANISM *mech)
 {
-	unsigned int ctxsize, devctxsize;
-	struct oc_sha_ctx *sc;
-
-	ctxsize = (sizeof(struct oc_sha_ctx) + 0x000F) & ~0x000F;
-	switch (mech->mechanism) {
-	case CKM_SHA_1:
-		devctxsize = sizeof(sha_context_t);
-		break;
+    unsigned int ctxsize, devctxsize;
+    struct oc_sha_ctx *sc;
+
+    ctxsize = (sizeof(struct oc_sha_ctx) + 0x000F) & ~0x000F;
+    switch (mech->mechanism) {
+    case CKM_SHA_1:
+        devctxsize = sizeof(sha_context_t);
+        break;
     case CKM_SHA224:
-	case CKM_SHA256:
-		devctxsize = sizeof(sha256_context_t);
-		break;
-	case CKM_SHA384:
-	case CKM_SHA512:
-		devctxsize = sizeof(sha512_context_t);
-		break;
-	default:
-		return CKR_MECHANISM_INVALID;
-	}
-
-	/* (re)alloc ctx in one memory area */
-	if (ctx->context)
-		free (ctx->context);
-	ctx->context_len = 0;
-	ctx->context = malloc(ctxsize + devctxsize);
-	if (ctx->context == NULL) {
-		TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
-		return CKR_HOST_MEMORY;
-	}
-	memset(ctx->context, 0, ctxsize + devctxsize);
-	ctx->context_len = ctxsize + devctxsize;
-	sc = (struct oc_sha_ctx *)ctx->context;
-	sc->dev_ctx_offs = ctxsize;
-
-	sc->message_part = SHA_MSG_PART_ONLY;
-	switch (mech->mechanism) {
-	case CKM_SHA_1:
-		sc->hash_len = SHA1_HASH_SIZE;
-		sc->hash_blksize = SHA1_BLOCK_SIZE;
-		break;
+    case CKM_SHA256:
+        devctxsize = sizeof(sha256_context_t);
+        break;
+    case CKM_SHA384:
+        devctxsize = sizeof(sha512_context_t);
+        break;
+    case CKM_SHA512:
```

Comment:
Is there a reason why the 2 cases CKM_SHA512 and CKM_SHA384 are now separated?

Reply:
That's what I've been waiting for! Since I've rebased many times those patch sets, I would expect some errors from the rebase. 
I will fix all the mentioned errors, just let me know when you are done with the review so I push the fixes.
Thanks!